### PR TITLE
feat(xast): add type definitions for xast

### PR DESCRIPTION
**Pull request by stefanprobst**
this adds type definitions for [xast](https://github.com/syntax-tree/xast).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/xast/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+     */
+    attributes?: Attributes;
+    children: Array<Element | Text | Comment | Instruction | Cdata>;
+}
+
+/**
+ * Information associated with an element.
+ */
+export interface Attributes {
+    [name: string]: string;
```
Comment (ChristianMurphy) [inline types/xast/index.d.ts:61]: are attribute values guaranteed to be string?
`hast` has a similar concept in properties, that supports a wide range of primitives.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c06fa19fdc843beedca332198de946928d24e18e/types/hast/index.d.ts#L72-L77

/cc @wooorm

Comment (stefanprobst) [reply types/xast/index.d.ts:61]: the `hast` docs have `typedef any PropertyValue`, while `xast` docs have `typedef string AttributeValue`, but yes would be good to confirm this

**Review by amcasey — CHANGES_REQUESTED** [review]
Please address the questions from the other reviewers.

File: types/xast/index.d.ts (hunk 1 above)
Comment (wooorm) [reply types/xast/index.d.ts:61]: Attribute values are indeed strings, unlike hast!

File: types/xast/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import { Parent as UnistParent, Literal as UnistLiteral, Node as UnistNode } from 'unist';
+
+export { UnistNode as Node };
+
+/**
+ * Node in xast containing other nodes.
+ * Its content is limited to only other xast content.
+ */
+export interface Parent extends UnistParent {
+    children: Array<Element | Text | Comment | DocType | Instruction | Cdata>;
```
Comment (wooorm) [inline types/xast/index.d.ts:19]: xast uses `Doctype`

File: types/xast/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * XML comment.
+ */
+export interface Comment extends Literal {
+    type: 'comment';
+}
+
+/**
+ * XML doctype.
+ */
+export interface DocType extends UnistNode {
```
Comment (wooorm) [inline types/xast/index.d.ts:81]:
```suggestion
export interface Doctype extends UnistNode {
```

File: types/xast/index.d.ts (hunk 2 above)
Comment (wooorm) [reply types/xast/index.d.ts:19]:
```suggestion
    children: Array<Element | Text | Comment | Doctype | Instruction | Cdata>;
```

**Review by amcasey — APPROVED** [review]
Thanks for confirming!

**Review by ChristianMurphy — APPROVED** [review]

File: types/xast/index.d.ts (hunk 1 above)
Comment (wooorm) [reply types/xast/index.d.ts:61]: Although, you might add `null` / `undefined` too, as those values *should* be handled by utilities

**Review by wooorm — APPROVED** [review]

**Review by wooorm — APPROVED** [review]

**Review by ChristianMurphy — APPROVED** [review]

**Review by amcasey — APPROVED** [review]

**Comment by amcasey** [comment]
@wooorm The bot just tagged me for the commits since my last sign-off.

**Comment by stefanprobst** [comment]
Ready to merge

**Pull request merged**